[WIP] Adds SQLAlchemy code for QuestionAnswerPairs

mfekadu opened the pull request:
## What's New?
* adds create & update functionality for Question & Answer data for the nimbus NLP pipeline

## Fixes #52 

## Type of change (pick-one)
- [ ] Bug fix (_non-breaking change which fixes an issue_)
- [x] New feature (_non-breaking change which adds functionality_)
- [ ] Breaking change (_fix or feature that would cause existing functionality to not work as expected_)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Yes see comment below

## Checklist (check-all-before-merge)
_formatting help: `- [x]` means "checked' and `- [ ]` means "unchecked"_

- [ ] I documented my code according to the [Google Python Style Guide][1]

- [ ] I ran `./build_docs.sh` and the docs look fine

- [ ] I ran `./type_check.sh` and got no errors

- [ ] I ran `./format.sh` because it automatically cleans my code for me 😄 

- [ ] I ran `./lint.sh` to check for what "format" missed

- [ ] I added my tests to the `/tests` directory

- [ ] I ran `./run_tests.sh` and all the tests pass

[1]: https://google.github.io/styleguide/pyguide.html

mfekadu committed:
add QuestionAnswerPair entity

mfekadu committed:
add save_question_answer_pair and include QAPair in database wrapper

mfekadu committed:
refactor with auto formatter & better comments & remove dead code

auto formatter = https://black.readthedocs.io/en/stable/

snekiam commented on Entity/QuestionAnswerPair.py:
```diff
@@ -0,0 +1,39 @@
+        """
+        D = self.__dict__
+        attributes = [
+            f"{k}={D.get(k)}" for k in self.__dir__() if not k.startswith("_")
```
Can you add a comment about what this line does? Even something as simple as "This creates a list of attributes that don't start with an '_'." could be helpful in the future I think.

mfekadu replied:
@snekiam  good point, I'll make another PR of refactorings for dead code, FIXMEs & TODOs

mfekadu commented:
# What's new
## Made the entity
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/Entity/QuestionAnswerPair.py#L1-L39

## Modified the `database_wrapper`
### import the entity
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L23

### add `self.QuestionAnswerPair`
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L253-L262

### do `__safe_create(self.QuestionAnswerPair)`
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L290-L305

### define `save_question_answer_pair`
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L332-L364

snekiam commented on database_wrapper.py:
```diff
@@ -247,72 +250,76 @@ def wrapper(*args, **kwargs):
         def __safe_create(SQLAlchemy_object):
             table_name = SQLAlchemy_object.__tablename__
-            print("creating {}...".format(table_name))
+            print(f"creating {table_name}...")
```
I'm not familiar with this syntax - is this how we want to do string formatting?

mfekadu replied:
Yes, `f-strings` are a thing https://realpython.com/python-f-strings/

They are super cool!

Thanks for your curiosity @snekiam

snekiam commented on database_wrapper.py:
```diff
@@ -322,12 +329,46 @@ def create_AudioSampleMetaData_table(self) -> None:
         self.AudioSampleMetaData.__table__.create(bind=self.engine)
 
+    def save_question_answer_pair(self, qa_dict: dict) -> bool:
+        """
+        Save the given question answer pair into the database.
+
+        Example input:
+        {
+            "can_we_answer": False,
+            "answer_type": AnswerType.other,
+            "question_format": "What is the meaning of life?",
+            "answer_format": "Dr. Fizzbuzz says the answer is sqrt(1764)"
+        }
+
+        Args:
+            qa_dict: a dictionary that corresponds to the fields in QuestionAnswerPair  # noqa
+
+        Raises:
+            BadDictionaryKeyError - ...
+            BadDictionaryValueError - ...
+
+        Returns:
+            True if all is good, else False
+        """
```
I love the long description of inputs and outputs! Should be very helpful for future maintenance.

mfekadu replied:
❤️

snekiam commented on database_wrapper.py:
```diff
@@ -377,48 +426,54 @@ def save_audio_sample_meta_data(self, formatted_data: dict) -> bool:
+            formatted_data["noiseLevel"] == "q"
+            or formatted_data["noiseLevel"] == "quiet"
+        ):
             metadata.noise_level = NoiseLevel.quiet
-        elif (formatted_data['noiseLevel'] == 'm' or
-              formatted_data['noiseLevel'] == 'medium'):
+        elif (
+            formatted_data["noiseLevel"] == "m"
+            or formatted_data["noiseLevel"] == "medium"
+        ):
             metadata.noise_level = NoiseLevel.medium
-        elif (formatted_data['noiseLevel'] == 'l' or
-              formatted_data['noiseLevel'] == 'loud'):
+        elif (
+            formatted_data["noiseLevel"] == "l"
+            or formatted_data["noiseLevel"] == "loud"
+        ):
```
I'm not sure if I like the multiline if statements - it seems like it takes up a lot of vertical space (I know this was the linter - maybe we should change some settings in the future? I don't think this should block the PR.)

mfekadu replied:
https://black.readthedocs.io

snekiam commented on database_wrapper.py:
```diff
@@ -607,10 +666,9 @@ def get_course_properties(self, courseName) -> List[str]:
-        props = self.get_property_from_entity(["*"],
-                                              "Courses",
-                                              condition_field="courseName",
-                                              condition_value=courseName)
+        props = self.get_property_from_entity(
+            ["*"], "Courses", condition_field="courseName", condition_value=courseName
+        )
```
Should this be commented out as indicated in the FIXME?

mfekadu replied:
@snekiam  good point, I'll make another PR of refactorings for dead code, FIXMEs & TODOs

snekiam reviewed (approved):
Looks good, maybe I need to read some linting docs though - I'm not familiar with the linter and some of it is probably best practice.

mfekadu committed:
manually tested and passed

(venv) ➜  api git:(mf-qa-pairs#52) ✗ ./database_wrapper.py
initialized database session
initialized NimbusMySQLAlchemy
creating Courses...
<Courses> already exists
creating AudioSampleMetaData...
<AudioSampleMetaData> already exists
creating QuestionAnswerPair...
<QuestionAnswerPair> created
NimbusMySQLAlchemy closed
(venv) ➜  api git:(mf-qa-pairs#52) ✗

mfekadu commented:
# How Has This Been Tested?

![Feb-23-2020 14-27-32](https://user-images.githubusercontent.com/11131250/75121421-aa1eca00-5648-11ea-9335-e6971cdaacb0.gif)

mfekadu committed:
Merge branch 'dev' into mf-qa-pairs#52
